projects/com.oracle.truffle.llvm.asm.amd64/src/com/oracle/truffle/llvm/asm/amd64/AsmFactory.java

```diff
@@ -0,0 +1,199 @@
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64AndlNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64DeclNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64ImmNode;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64InclNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64MovlNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64NotlNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64OrlNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64SallNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64SarlNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64ShllNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64ShrlNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64SublNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.asm.LLVMAMD64XorlNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.base.integers.LLVMI32Node;
+import com.oracle.truffle.llvm.nodes.impl.func.LLVMArgNodeFactory.LLVMI32ArgNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.func.LLVMInlineAssemblyRootNode;
+import com.oracle.truffle.llvm.nodes.impl.others.LLVMUnsupportedInlineAssemblerNode.LLVMI32UnsupportedInlineAssemblerNode;
+import com.oracle.truffle.llvm.nodes.impl.vars.LLVMReadNodeFactory.LLVMI32ReadNodeGen;
+import com.oracle.truffle.llvm.nodes.impl.vars.LLVMWriteNodeFactory.LLVMWriteI32NodeGen;
+import com.oracle.truffle.llvm.parser.LLVMBaseType;
+
+public class AsmFactory {
+
+    private static final int HEX = 16;
+
+    private FrameDescriptor frameDescriptor;
+    private List<LLVMNode> statements;
+    private List<LLVMNode> arguments;
+    private List<String> registers;
+    private LLVMExpressionNode result;
+    private String asmFlags;
+
+    public AsmFactory(String asmFlags) {
+        this.asmFlags = asmFlags;
+        this.frameDescriptor = new FrameDescriptor();
+        this.statements = new ArrayList<>();
+        this.arguments = new ArrayList<>();
+        this.registers = new ArrayList<>();
+        this.registers.add("-");
+    }
+
+    public LLVMInlineAssemblyRootNode finishInline() {
+        getArguments();
+        return new LLVMInlineAssemblyRootNode(null, frameDescriptor, statements.toArray(new LLVMNode[statements.size()]), arguments, result);
+    }
+
+    public void createBinaryOperation(String operation, String left, String right) {
+        FrameSlot leftSlot = frameDescriptor.findFrameSlot(left);
+        FrameSlot rightSlot = frameDescriptor.findFrameSlot(right);
+        LLVMI32Node leftNode = (leftSlot != null) ? LLVMI32ReadNodeGen.create(leftSlot) : getImmediateNode(left);
+        LLVMI32Node rightNode = LLVMI32ReadNodeGen.create(rightSlot);
+        this.result = LLVMI32ReadNodeGen.create(rightSlot);
+        switch (operation) {
+            case "addl":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64AddlNodeGen.create(leftNode, rightNode), rightSlot));
```

Comment:
Could you assign the binary operation to a variable and use `this.statements.add(LLVMWriteI32NodeGen.create(...))` only once?